src/lua/functions/core/libs/kv_functions.cpp

```diff
@@ -0,0 +1,138 @@
+/**
+ * Canary - A free and open-source MMORPG server emulator
+ * Copyright (©) 2019-2022 OpenTibiaBR <opentibiabr@outlook.com>
+ * Repository: https://github.com/opentibiabr/canary
+ * License: https://github.com/opentibiabr/canary/blob/main/LICENSE
+ * Contributors: https://github.com/opentibiabr/canary/graphs/contributors
+ * Website: https://docs.opentibiabr.com/
+ */
+
+#include "pch.hpp"
+
+#include <variant>
+
+#include "kv/kv.hpp"
+#include "lua/functions/core/libs/kv_functions.hpp"
+#include "lua/scripts/lua_environment.hpp"
+
+int KVFunctions::luaKVScoped(lua_State* L) {
+	// KV.scoped(key)
+	auto key = getString(L, -1);
+	auto scopedKV = g_kv().scoped(key);
+	pushUserdata<KVStore::ScopedKV>(L, scopedKV);
+	setMetatable(L, -1, "KVStore::ScopedKV");
+	return 1;
+}
+
+int KVFunctions::luaKVSet(lua_State* L) {
+	// KV.set(key, value) | scopedKV:set(key, value)
+	auto key = getString(L, -2);
+	auto valueWrapper = getValueWrapper(L);
+
+	if (!valueWrapper) {
+		g_logger().warn("[{}] invalid param type", __FUNCTION__);
+		pushBoolean(L, false);
+		return 1;
+	}
+
+	if (isUserdata(L, 1)) {
+		auto scopedKV = getUserdata<KVStore::ScopedKV>(L, 1);
+		scopedKV->set(key, valueWrapper.value());
+		pushBoolean(L, true);
+		return 1;
+	}
+
+	g_kv().set(key, valueWrapper.value());
+	pushBoolean(L, true);
+	return 1;
+}
+
+int KVFunctions::luaKVGet(lua_State* L) {
+	// KV.get(key) | scopedKV:get(key)
+	std::optional<ValueWrapper> valueWrapper;
+	auto key = getString(L, -1);
+	if (isUserdata(L, 1)) {
+		auto scopedKV = getUserdata<KVStore::ScopedKV>(L, 1);
+		valueWrapper = scopedKV->get(key);
+	} else {
+		valueWrapper = g_kv().get(key);
+	}
+
+	if (valueWrapper) {
+		pushValueWrapper(L, *valueWrapper);
+	} else {
+		lua_pushnil(L);
+	}
+	return 1;
+}
+
+std::optional<ValueWrapper> KVFunctions::getValueWrapper(lua_State* L) {
+	if (isString(L, -1)) {
+		return getString(L, -1);
+	} else if (isNumber(L, -1)) {
+		return getNumber<double>(L, -1);
+	} else if (isBoolean(L, -1)) {
+		return getBoolean(L, -1);
+	} else if (isTable(L, -1)) {
+		if (lua_objlen(L, -1) > 0) {
+			ArrayType array;
+			for (int i = 1; i <= lua_objlen(L, -1); ++i) {
+				lua_rawgeti(L, -1, i);
+				auto value = getValueWrapper(L);
+				if (!value) {
+					g_logger().warn("[{}] invalid param type", __FUNCTION__);
+					return std::nullopt;
+				}
+				array.push_back(value.value());
+				lua_pop(L, 1);
+			}
+			return ValueWrapper(array);
+		} else {
+			MapType map;
+			lua_pushnil(L);
+			while (lua_next(L, -2) != 0) {
+				auto value = getValueWrapper(L);
+				if (!value) {
+					g_logger().warn("[{}] invalid param type", __FUNCTION__);
+					return std::nullopt;
+				}
+				map[getString(L, -2)] = std::make_shared<ValueWrapper>(value.value());
+				lua_pop(L, 1);
+			}
+			return ValueWrapper(map);
+		}
+	} else if (isNil(L, -1)) {
+		return nullptr;
+	} else {
+		g_logger().warn("[{}] invalid param type", __FUNCTION__);
+		return std::nullopt;
+	}
```

Comment:
This if/else chain could be simplified since you're always returning the elses are not needed.